### PR TITLE
eval on intermediate checkpoints

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR allows for eval on intermediate checkpoints. Shows up in the leaderboard like this:

Checked with 

image
image

```
python mason.py \
    --cluster ai2/jupiter-cirrascale-2 --image nathanl/open_instruct_auto --pure_docker_mode \
    --workspace ai2/tulu-3-dev \
    --priority urgent \
    --preemptible \
    --image costah/open_instruct_ppo_ray_dev \
    --budget ai2/allennlp \
    --gpus 8 -- WANDB_TAGS=beta_ablation python open_instruct/ppo_vllm_thread_ray_gtrl_weka.py \
    --exp_name ppo_ray_β_0.03 \
    --dataset_mixer '{"ai2-adapt-dev/gsm8k_ground_truth": 1.0, "trl-internal-testing/tldr-preference-sft-trl-style": 1.0}' \
    --dataset_train_splits train \
    --dataset_eval_mixer '{"ai2-adapt-dev/gsm8k_math_ground_truth": 1.0}' \
    --dataset_eval_splits test \
    --max_token_length 2048 \
    --max_prompt_token_length 2048 \
    --response_length 1024 \
    --model_name_or_path allenai/open_instruct_dev \
    --model_revision dpo_tune___model__42__1729311739 \
    --reward_model_path allenai/open_instruct_dev \
    --reward_model_revision reward_modeling__1__1729382759 \
    --non_stop_penalty \
    --stop_token eos \
    --temperature 1.0 \
    --ground_truths_key ground_truth \
    --chat_template tulu \
    --sft_messages_key messages \
    --learning_rate 3e-7 \
    --total_episodes 2024 \
    --penalty_reward_value -10.0 \
    --deepspeed_stage 3 \
    --per_device_train_batch_size 4 \
    --local_rollout_forward_batch_size 8 \
    --local_mini_batch_size 32 \
    --local_rollout_batch_size 32 \
    --actor_num_gpus_per_node 7 \
    --vllm_tensor_parallel_size 1 \
    --beta 0.03 \
    --apply_verifiable_reward true \
    --output_dir /weka/oe-adapt-default/costah/models/ppo_mixing \
    --seed 3 \
    --num_evals 3 \
    --reward_model_multiplier 0.1 \
    --gradient_checkpointing \
    --no_try_launch_beaker_eval_jobs \
    --try_launch_beaker_eval_jobs_on_weka \
    --oe_eval_tasks "gsm8k::tulu" \
    --save_freq 4 \
    --with_tracking \
    --push_to_hub
```

### CODE REVIEW

File: open_instruct/model_utils.py [hunk 1]
```diff
@@ -227,6 +227,9 @@ def apply_verifiable_reward(
     rewards = []
     for prediction, ground_truth, dataset in zip(decoded_responses, ground_truths, datasets):
         verified = False
+        if ground_truth is None:
```
Comment (vwxyzjn) [inline open_instruct/model_utils.py:230]: This makes the code work with mixing different dataset.

File: open_instruct/ppo_vllm_thread_ray_gtrl_weka.py [hunk 2]
```diff
@@ -0,0 +1,1780 @@
...
+
+    # create the dataset
+    dataset_dict = DatasetDict()
+    dataset_processor = SFTGroundTruthDatasetProcessor(tokenizer=tokenizer, config=dataset_config)
+    if len(args.dataset_train_splits) != len(args.dataset_mixer_dict) and len(args.dataset_train_splits) == 1:
+        args.dataset_train_splits = [args.dataset_train_splits[0]] * len(args.dataset_mixer_dict)
+        print(f"Dataset splits not provided for all datasets. Using the same {args.dataset_train_splits[0]} split for all datasets.")
+    if len(args.dataset_eval_splits) != len(args.dataset_eval_mixer_dict) and len(args.dataset_eval_splits) == 1:
+        args.dataset_eval_splits = [args.dataset_eval_splits[0]] * len(args.dataset_eval_mixer_dict)
+        print(f"Dataset splits not provided for all datasets. Using the same {args.dataset_eval_splits[0]} split for all datasets.")
```
Comment (vwxyzjn) [inline open_instruct/ppo_vllm_thread_ray_gtrl_weka.py:1598]: Btw @hamishivi this appears to be a bug in other implementations: if I mix dataset without specifying multiple train splits, the mix is prob not successful

Comment (hamishivi) [reply open_instruct/ppo_vllm_thread_ray_gtrl_weka.py:1598]: good catch. This probably affected me, although it doesnt quite explain why just MATH data alone isnt working well :(

**Review by hamishivi — APPROVED** [review]
I honestly dont mind merging the gtrl_weka script with the old ray_gtrl script, but don't feel super strongly about it.

**Pull request merged**